src/main/java/de/hysky/skyblocker/skyblock/dungeon/DungeonScore.java

```diff
@@ -1,76 +1,369 @@
+		return matcher != null ? Integer.parseInt(matcher.group("crypts")) : 0;
+	}
+
+	public static boolean hasSpiritPet(String name) {
+		return SpiritPetCache.computeIfAbsent(name, k -> {
+			String playeruuid = ApiUtils.name2Uuid(name);
+			try (Http.ApiResponse response = Http.sendHypixelRequest("skyblock/profiles", "?uuid=" + playeruuid)) {
+				if (!response.ok()) throw new IllegalStateException("Failed to get profile uuid for player " + name + "! Response: " + response.content());
+				JsonObject responseJson = JsonParser.parseString(response.content()).getAsJsonObject();
+
+				JsonObject player = StreamSupport.stream(responseJson.getAsJsonArray("profiles").spliterator(), false)
+						.map(JsonElement::getAsJsonObject)
+						.filter(profile -> profile.getAsJsonPrimitive("selected").getAsBoolean())
+						.findFirst()
+						.orElseThrow(() -> new IllegalStateException("No selected profile found!?"))
+						.getAsJsonObject("members").entrySet().stream()
+						.filter(entry -> entry.getKey().equals(playeruuid))
+						.map(Map.Entry::getValue)
+						.map(JsonElement::getAsJsonObject)
+						.findFirst()
+						.orElseThrow(() -> new IllegalStateException("Player somehow not found inside their own profile!"));
+
+				for (JsonElement element : player.getAsJsonObject("pets_data").getAsJsonArray("pets")) {
+					if (!element.getAsJsonObject().get("type").getAsString().equals("SPIRIT")) continue;
+					if (!element.getAsJsonObject().get("tier").getAsString().equals("LEGENDARY")) continue;
+
+					return true;
+				}
+			} catch (Exception e) {
+				LOGGER.error("[Skyblocker] Spirit pet lookup by name failed! Name: {} - Cause: {}", name, e.getMessage());
+			}
+			return false;
+		});
+	}
+
+	private static void checkMessageForDeaths(String message) {
+		if (!message.startsWith("\u2620", 1)) return;
+		Matcher matcher = DEATHS_PATTERN.matcher(message);
+		if (!matcher.matches()) return;
+		deathCount++;
+		if (deathCount > 1) return;
+		final String whoDied = matcher.group("whodied").transform(s -> {
+			if (s.equals("You")) return MinecraftClient.getInstance().player.getName().getString(); //This will be wrong if the dead player is called 'You' but that's unlikely
+			else return s;
+		});
+		CompletableFuture.supplyAsync(() -> hasSpiritPet(whoDied))
+				.thenAccept(hasSpiritPet -> firstDeathHasSpiritPet = hasSpiritPet);
+	}
+
+	private static void checkMessageForWatcher(String message) {
+		if (message.equals("[BOSS] The Watcher: You have proven yourself. You may pass.")) bloodRoomCompleted = true;
+	}
+
+	public static void setCurrentFloor() {
+		for (String sidebarLine : Utils.STRING_SCOREBOARD) {
+			Matcher floorMatcher = FLOOR_PATTERN.matcher(sidebarLine);
+			if (!floorMatcher.matches()) continue;
+			currentFloor = floorMatcher.group("floor");
+			return;
+		}
+		LOGGER.error("[Skyblocker] Floor pattern doesn't match!");
+	}
+
+	public static int getScore() {
+		return score;
+	}
+
+	public static boolean isDungeonStarted() {
```

Comment:
This method might make more sense to be in the DungeonManager class but not that big of a deal